test/Gen.hs

```diff
@@ -125,52 +143,72 @@ class Forall g f | g -> f, f -> g where
 instance Forall (Rec '[]) (PropertyT IO ()) where
   forall' Nil = id
 
-instance (Forall (Rec gs) b, Show a) => Forall (Rec (Gen a ': gs)) (a -> b) where
+instance (Forall (Rec gs) b) => Forall (Rec (Gen a ': gs)) (a -> b) where
   forall' (g :. gs) f = do
-    a <- Hedgehog.forAll g
-    forall' gs (f a)
+    (a, labels) <- Hedgehog.forAll (runWriterT (runGen g))
+    traverse_ Hedgehog.label labels
+    forall' gs (f (runTerm a))
 
 
-data With a = With' { _showWith :: Int -> ShowS, getWith :: a }
-  deriving (Functor)
+showing :: Show a => a -> Term a
+showing = Pure . flip showsPrec <*> id
+
+showingFn :: (Show a, Show b) => Fn.Fn a b -> Term (a -> b)
+showingFn = Pure . flip showsPrec <*> Fn.apply
 
-instance Eq a => Eq (With a) where
-  (==) = (==) `on` getWith
+atom :: String -> a -> Gen a
+atom s = Gen . pure . Pure (const (showString s))
 
-instance Applicative With where
-  pure = atom "_"
-  With' sf f <*> With' sa a = With' (\ d -> showParen (d > 10) (sf 10 . showString " " . sa 11)) (f a)
+label :: String -> a -> Gen a
+label s = addLabel s . atom s
 
-instance Show (With a) where
-  showsPrec d (With' s _) = s d
+infixL :: Int -> String -> (a -> b -> c) -> Gen (a -> b -> c)
+infixL p s f = Gen (pure (InfixL p s f))
 
-showing :: Show a => a -> With a
-showing = With' . flip showsPrec <*> id
+infixR :: Int -> String -> (a -> b -> c) -> Gen (a -> b -> c)
+infixR p s f = Gen (pure (InfixR p s f))
 
-showingFn :: (Show a, Show b) => Fn a b -> With (a -> b)
-showingFn = With' . flip showsPrec <*> apply
+addLabel :: String -> Gen a -> Gen a
+addLabel s = Gen . (>>= \ a -> a <$ tell (Set.singleton (fromString s))) . runGen
 
-atom :: String -> a -> With a
-atom s = With' (\ _ -> showString s)
 
-liftWith :: String -> (a -> b) -> With a -> With b
-liftWith s w a = atom s w <*> a
+data Term a where
+  Pure :: (Int -> ShowS) -> a -> Term a
+  InfixL :: Int -> String -> (a -> b -> c) -> Term (a -> b -> c)
+  InfixR :: Int -> String -> (a -> b -> c) -> Term (a -> b -> c)
+  (:<*>) :: Term (a -> b) -> Term a -> Term b
```

Comment:
I decided this all wasn’t over-engineered enough, so I rewrote `With` as a deep embedding, et voilà, `Term`.

The main takeaway is that this allows us to do smarter partial application—basically, operator sections with correct precedence. It also makes it easier and more concise to combine labels, resulting in much briefer, clearer code. None of those `liftWith*`s cluttering things up!